### DISCUSSION TITLE
MultiApp transfer material property

### DISCUSSION THREAD

**Discussion by tqcthai** [idaholab/moose#27586] [Q&A General] [answered]
Dear all,

I start using Multi-app but fail in running it. In the main-app, I try to push the information of 'elastic_energy' which is computed from the stress function to the sub-app, and pull the 'fracture_energy' from it.

The error I encounter is `sub_app0: Material property 'elastic_energy', requested by 'fracture_driving_energy' is not defined on block 1
`

Main-app
```
[AuxVariables]
  [elastic_energy]
   order = FIRST
   family = MONOMIAL
  []
[]
[AuxKernels]
 [elastic_energy]
  type = MaterialRealAux
  variable = elastic_energy
  property = elastic_energy
 []
[]


[Materials]
  [elasticity_tensor]
    type = ComputeElasticityTensor
    C_ijkl = '136.5e3 70.3e3'
    fill_method = symmetric_isotropic
  []
  [compute_stress]
    c = c
    type = ComputeJ2PlasticityPFStress
    E_name = 'elastic_energy'
    F_name = 'local_fracture_energy'
  []
[]


[MultiApps]
  [sub_app]
    type = TransientMultiApp
    positions = '0 0 0'
    input_files = 'PF.i'
  []
[]

[Transfers]
  [pull_v]
    type = MultiAppProjectionTransfer
    from_multi_app = sub_app
    source_variable = c
    variable = local_fracture_energy
  []
  [push_u]
    type = MultiAppProjectionTransfer
    to_multi_app = sub_app
    source_variable = displacement
    variable = elastic_energy
  []
[]
```

Sub-app
```

[AuxVariables]
  [local_fracture_energy]
    order = FIRST
    family = MONOMIAL
  []
  [elastic_energy]
  []
[]

[AuxKernels]
  [local_fracture_energy]
    type = MaterialRealAux
    variable = local_fracture_energy
    property = local_fracture_energy
  []
[]


[Materials]
  [local_fracture_energy]
    type = DerivativeParsedMaterial
    property_name = local_fracture_energy
    coupled_variables = 'c'
    material_property_names = 'gc_prop l'
    expression = 'c^2 * gc_prop / 2 / l'
    derivative_order = 2
  []
  [fracture_driving_energy]
    type = DerivativeSumMaterial
    coupled_variables = c
    sum_materials = 'elastic_energy local_fracture_energy'
    derivative_order = 2
    property_name = F
  []
[]
```
Hope to get your help.

**Comment by GiudGiud** [comment]
Hello

The first issue is this block
```
  [pull_v]
    type = MultiAppProjectionTransfer
    from_multi_app = sub_app
    source_variable = c
    variable = local_fracture_energy
  []
```
which refers to a 'local_fracture_energy' variable when it looks like it s only a material property.
You can use the material property output (on each material) to trigger the output of the property to a variable. 

For debugging other material property issues, use this:
```
[Debug]
  show_material_props = true
[]
```

**Reply by tqcthai** [reply]
That's true. But I think there is also an error for passing the 'elastic energy' from the main-app to the sub-app. The error I get is in the sub_app : "sub_app0: Material property 'elastic_energy', requested by 'fracture_driving_energy' is not defined on block 1". Is the way I define 'elastic_energy' as an AuxVariable correct to pass it as
```
[push_u]
    type = MultiAppProjectionTransfer
    to_multi_app = sub_app
    source_variable = displacement
    variable = elastic_energy
  []
```
?

**Reply by GiudGiud** [reply, marked as answer]
the object returning the error `'fracture_driving_energy` is not the transfer. 
it's a DerivativeSumMaterial
and it's telling you there's no elastic_energy material property. Only a variable is available.

So either create a material property from the variable (using a VarMaterial if that exists, or similar)
or use a variable in the DerivativeSumMaterial

**Reply by tqcthai** [reply]
Thanks @GiudGiud. I made it work.

One more question: In case I am using the Newton solver for both the main-problem and the sub-problem, as far as I can see, the Multiapp solves the equation of the main-problem by iterating until the convergence criteria are fulfilled. Then it transfers the data to the sub-problem and solves it in the same way, i.e., by iterating and checking for convergence.

How can I set up a framework like this:
1. Run the first Newton iteration for the main-problem.
2. Transfer the data from the main problem to the sub-problem.
3. Run the first Newton iteration for the sub-problem.
4. Check for convergence and repeat step 1 for the next Newton iteration until both solutions converge, then move to the next loading/time step.

Another question: Is transferring data between different apps expensive?

**Reply by GiudGiud** [reply]
hello

> Run the first Newton iteration for the main-problem.

- If you mean WITHIN the newton solve
you can do this using transfers and multiapps (both) executed on (set the `execute_on` parameter) on NONLINEAR. 

> Another question: Is transferring data between different apps expensive?

should not be as expensive as solving in the vast majority of cases.

- If you mean fixed point iterations OUTSIDE the newton solve, WITHIN one loading step:

in the execution block, set the fixed_point_max_its parameter to trigger fixed point iterations outside of the nonlinear solve
